compiler/rustc_interface/src/interface.rs

```diff
@@ -123,13 +124,89 @@ pub fn parse_cfgspecs(cfgspecs: Vec<String>) -> FxHashSet<(String, Option<String
+            names_valid: Default::default(),
+            values_checked: Default::default(),
+            values_valid: Default::default(),
+        };
+        for s in specs {
+            let sess = ParseSess::with_silent_emitter();
+            let filename = FileName::cfg_spec_source_code(&s);
+            let mut parser = new_parser_from_source_str(&sess, filename, s.to_string());
+
+            macro_rules! error {
+                ($reason: expr) => {
+                    early_error(
+                        ErrorOutputType::default(),
+                        &format!(
+                            concat!("invalid `--check-cfg` argument: `{}` (", $reason, ")"),
+                            s
+                        ),
+                    );
+                };
+            }
+
+            match &mut parser.parse_meta_item() {
+                Ok(meta_item) if parser.token == token::Eof => {
+                    if let Some(args) = meta_item.meta_item_list() {
+                        if meta_item.has_name(sym::names) {
+                            cfg.names_checked = true;
+                            for arg in args {
+                                if arg.is_word() && arg.ident().is_some() {
+                                    let ident = arg.ident().expect("multi-segment cfg key");
+                                    cfg.names_valid.insert(ident.name.to_string());
+                                } else {
+                                    error!("`names()` arguments must be simple identifers");
+                                }
+                            }
+                            continue;
+                        } else if meta_item.has_name(sym::values) {
+                            if let Some((name, values)) = args.split_first() {
+                                if name.is_word() && name.ident().is_some() {
+                                    let ident = name.ident().expect("multi-segment cfg key");
+                                    cfg.values_checked.insert(ident.to_string());
+                                    for val in values {
+                                        if let Some(lit) = val.literal() {
+                                            if let LitKind::Str(s, _) = lit.kind {
+                                                cfg.values_valid
+                                                    .insert((ident.to_string(), s.to_string()));
+                                                continue;
+                                            }
+                                        }
+                                        error!("`values()` arguments must be string literals");
+                                    }
+                                } else {
+                                    error!("`values()` first argument must be a simple identifer");
+                                }
+                                continue;
+                            }
```

Comment:
Need to produce an error on `--check-cfg=values()` (with a test case).

Reply:
Ah, ok, the error *is* reported below.
All this control flow with `continue`s in this function is hard to read, perhaps you could use `continue`s with labels at least?

Reply:
But the test cases are needed anyway, for all erroneous conditions that can be reported in this function.